src/passthrough/file_handle.rs

```diff
@@ -9,36 +9,153 @@ use std::fmt::{Debug, Formatter};
 use std::fs::File;
 use std::io;
 use std::os::unix::io::{AsRawFd, FromRawFd, RawFd};
-use std::ptr;
 use std::sync::{RwLock, RwLockReadGuard, RwLockWriteGuard};
 
+use vm_memory::ByteValued;
+use vmm_sys_util::fam::{FamStruct, FamStructWrapper};
+
 use crate::passthrough::PassthroughFs;
 
 /// An arbitrary maximum size for CFileHandle::f_handle.
 ///
 /// According to Linux ABI, struct file_handle has a flexible array member 'f_handle', but it's
 /// hard-coded here for simplicity.
-//pub const MAX_HANDLE_SZ: usize = 128;
+pub const MAX_HANDLE_SIZE: usize = 128;
+
+/// Dynamically allocated array.
+#[derive(Default)]
+#[repr(C)]
+pub struct __IncompleteArrayField<T>(::std::marker::PhantomData<T>, [T; 0]);
+impl<T> __IncompleteArrayField<T> {
+    #[inline]
+    pub unsafe fn as_ptr(&self) -> *const T {
+        self as *const __IncompleteArrayField<T> as *const T
+    }
+    #[inline]
+    pub unsafe fn as_mut_ptr(&mut self) -> *mut T {
+        self as *mut __IncompleteArrayField<T> as *mut T
+    }
+    #[inline]
+    pub unsafe fn as_slice(&self, len: usize) -> &[T] {
+        ::std::slice::from_raw_parts(self.as_ptr(), len)
+    }
+    #[inline]
+    pub unsafe fn as_mut_slice(&mut self, len: usize) -> &mut [T] {
+        ::std::slice::from_raw_parts_mut(self.as_mut_ptr(), len)
+    }
+}
 
-#[derive(Clone, Copy)]
+/// The structure to transfer file_handle struct between user space and kernel space.
+/// ```c
+/// struct file_handle {
+///     __u32 handle_bytes;
+///     int handle_type;
+///     /* file identifier */
+///     unsigned char f_handle[];
+/// }
+/// ```
+#[derive(Default)]
 #[repr(C)]
+pub struct MockCFileHandle {
+    pub len: u32,
+    pub padding: u32,
+    pub entries: __IncompleteArrayField<libc::c_char>,
+}
+
+unsafe impl FamStruct for MockCFileHandle {
+    type Entry = libc::c_char;
+
+    fn len(&self) -> usize {
+        self.len as usize
+    }
+
+    fn set_len(&mut self, len: usize) {
+        self.len = len as u32
+    }
+
+    fn max_len() -> usize {
+        // 8 means bytes for handle_bytes and handle_type
+        8 + MAX_HANDLE_SIZE
+    }
+
+    fn as_slice(&self) -> &[u8] {
+        let len = self.len();
+        unsafe { self.entries.as_slice(len) }
+    }
+
+    fn as_mut_slice(&mut self) -> &mut [Self::Entry] {
+        let len = self.len();
+        unsafe { self.entries.as_mut_slice(len) }
+    }
+}
+
+type MockCFileHandleWrapper = FamStructWrapper<MockCFileHandle>;
+
+#[derive(Clone)]
 struct CFileHandle {
     // Size of f_handle [in, out]
-    handle_bytes: libc::c_uint,
+    handle_bytes: u32,
     // Handle type [out]
-    handle_type: libc::c_int,
+    handle_type: i32,
     // File identifier (sized by caller) [out]
-    f_handle: *mut libc::c_char,
+    f_handle: Vec<u8>,
```

Comment:
```suggestion
 type  CFileHandle = FamStructWrapper<CFileHandleInner>;
 
 pub struct CFileHandlerInner {
     handle_bytes: u32,
     handle_type: i32,
     f_handle: __IncompleteArrayField<libc::c_char>
 }
```

Reply:
Thanks, @wllenyj! As CFileHandle needs to implement some traits, like Ord, PartialOrd, I use a CFileHandleWrapper instead, which is used as a memory of CFileHandle struct.